### PR TITLE
Training module using SocketIO-Flask

### PR DESCRIPTION
**Pull request by YasheshSavani**
Description:

This PR will implement training module in sense studio by using `Flask-Socketio` for communication between web-page and back-end to show output on the implemented terminal.

### CODE REVIEW

File: tools/sense_studio/static/training.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Ref: https://github.com/miguelgrinberg/Flask-SocketIO-Chat/blob/master/app/templates/chat.html
+var socket;
+$(document).ready(function () {
+
+    let project = $('#project').val();
+    let output_folder = $('#output_folder').val();
+
+    socket = io.connect('http://' + document.domain + ':' + location.port + '/train-model');
```
Comment (corneliusboehm) [inline tools/sense_studio/static/training.js:8]: In our current case the `http://` is not even needed

File: tools/sense_studio/training.py [hunk 2]
```diff
@@ -40,68 +33,88 @@ def train_model():
...
-#
-#     if os.path.exists(os.path.join(img_path, 'confusion_matrix.png')):
-#         return send_from_directory(img_path, 'confusion_matrix.png', as_attachment=True)
-#     return None
+@training_bp.route('/confusion-matrix/<string:project>', methods=['GET'])
+@training_bp.route('/confusion-matrix/<string:project>/<string:output_folder>', methods=['GET'])
+def confusion_matrix(project, output_folder):
+    path = utils.lookup_project_path(project)
+    img_path = os.path.join(path, output_folder, 'checkpoints')
+    return send_from_directory(img_path, filename='confusion_matrix.png', as_attachment=True)
```
Comment (corneliusboehm) [inline tools/sense_studio/training.py:120]: I think this is not used

File: tools/sense_studio/training.py [hunk 3]
```diff
@@ -40,68 +33,88 @@ def train_model():
...
+    global train_process
+    if train_process:
+        train_process.terminate()
+        train_process = None
+        message = "Training Cancelled."
+    return render_template('training.html', project=project, path=path, models=utils.BACKBONE_MODELS,
+                           is_disabled=False, message=message, output_folder=output_folder)
+
+
+@socketio.on('training_logs', namespace='/train-model')
```
Comment (corneliusboehm) [inline tools/sense_studio/training.py:84]: Can we give this message another name? I took me some time to understand that this is the "connect" message from the client

File: tools/sense_studio/training.py (hunk 2 above)
Comment (YasheshSavani) [reply tools/sense_studio/training.py:120]: This is being used in `training.js` where the `img` tag is being created.

Comment (YasheshSavani) [reply tools/sense_studio/training.py:120]: In that src will be given a path, which will render image using this method.

Comment (corneliusboehm) [reply tools/sense_studio/training.py:120]: Oh, I saw the `url_for` call a few lines above just now

File: tools/sense_studio/__init__.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    from tools.sense_studio.annotation import annotation_bp
+    app.register_blueprint(annotation_bp, url_prefix='/annotation')
+
+    from tools.sense_studio.video_recording import video_recording_bp
+    app.register_blueprint(video_recording_bp, url_prefix='/video-recording')
+
+    from tools.sense_studio.training import training_bp
+    app.register_blueprint(training_bp, url_prefix='/training')
+
+    socketio.init_app(app, cors_allowed_origins="*")
```
[the remark this answers is not in the document]
Comment (YasheshSavani) [reply tools/sense_studio/__init__.py:22]: I'll check it again why I added that

File: tools/sense_studio/training.py (hunk 3 above)
Comment (YasheshSavani) [reply tools/sense_studio/training.py:84]: Technically, it's an event name `training_logs` likewise `connect`. This one method works as connection, logs retriever, and at last disconnect sender.

Comment (YasheshSavani) [reply tools/sense_studio/training.py:84]: yes, if it bothers more we can think about something else

**Review by corneliusboehm — APPROVED** [review]

**Pull request merged**